src/nupic/algorithms/AnomalyLikelihood.hpp

```diff
@@ -0,0 +1,171 @@
+    namespace anomaly {
+
+struct DistributionParams {
+  DistributionParams(std::string name, Real32 mean, Real32 variance, Real32 stdev) :
+    name(name),mean(mean), variance(variance), stdev(stdev) {}
+  std::string name;
+  Real32 mean;
+  Real32 variance;
+  Real32 stdev;
+};
+
+class AnomalyLikelihood {
+
+  public:
+    AnomalyLikelihood(UInt learningPeriod=288, UInt estimationSamples=100, UInt historicWindowSize=8640, UInt reestimationPeriod=100, UInt aggregationWindow=10);
+
+    Real anomalyProbability(Real rawValue, Real anomalyScore, int timestamp=-1);
+
+  private:
+    //methods:
+    vector<Real> estimateAnomalyLikelihoods(vector<Real> anomalyScores, UInt skipRecords=0, UInt verbosity=0);
+    vector<Real>  updateAnomalyLikelihoods(vector<Real> anomalyScores, UInt verbosity=0);
+    Real32 tailProbability(Real32 x) const;
+    DistributionParams estimateNormal(vector<Real> sampleData, bool performLowerBoundCheck=true);
+    DistributionParams nullDistribution() const;
+
+    //private static methods
+static Real  computeLogLikelihood(Real likelihood)  {
+  /**
+    Compute a log scale representation of the likelihood value. Since the
+    likelihood computations return low probabilities that often go into four 9's
+    or five 9's, a log value is more useful for visualization, thresholding,
+    etc.
+   **/
+    // The log formula is:
+    //     Math.log(1.0000000001 - likelihood) / Math.log(1.0 - 0.9999999999)
+    return log(1.0000000001f - likelihood) / -23.02585084720009f;
+}
+
+static UInt calcSkipRecords_(UInt numIngested, UInt windowSize, UInt learningPeriod)  {
+    /** Return the value of skipRecords for passing to estimateAnomalyLikelihoods
+
+    If `windowSize` is very large (bigger than the amount of data) then this
+    could just return `learningPeriod`. But when some values have fallen out of
+    the historical sliding window of anomaly records, then we have to take those
+    into account as well so we return the `learningPeriod` minus the number
+    shifted out.
+
+    @param numIngested - (int) number of data points that have been added to the
+      sliding window of historical data points.
+    @param windowSize - (int) size of sliding window of historical data points.
+    @param learningPeriod - (int) the number of iterations required for the
+      algorithm to learn the basic patterns in the dataset and for the anomaly
+      score to 'settle down'.
+    **/
+    int diff = numIngested - (int)windowSize;
+    UInt numShiftedOut = max(0, diff);
+    return min(numIngested, max((UInt)0, learningPeriod - numShiftedOut));
+}
+
+static Real compute_mean(vector<Real> v)  { //TODO do we have a (more comp. stable) implementation of mean/variance?
+    Real sum = std::accumulate(v.begin(), v.end(), 0.0);
+    return sum / v.size();
+}
+
+static Real compute_var(vector<Real> v, Real mean)  {
+    Real sq_sum = std::inner_product(v.begin(), v.end(), v.begin(), 0.0);
+    return sq_sum / v.size() - mean * mean;
```

Comment:
I think it'd be better to use explicit parentheses to show order of operations. This code works, but it will be more readable and less scary with parens.